src/main/scala/firrtl/passes/MemUtils.scala

```diff
@@ -0,0 +1,185 @@
+import firrtl.Utils._
+import firrtl.Mappers._
+import firrtl.PrimOps._
+import firrtl.Annotations._
+import firrtl.WrappedExpression._
+
+object toBits {
+  def apply(e: Expression): Expression = {
+    e match {
+      case ex: WRef => hiercat(ex,ex.tpe)
+      case ex: WSubField => hiercat(ex,ex.tpe)
+      case ex: WSubIndex => hiercat(ex,ex.tpe)
+      case t => error("Invalid operand expression for toBits!")
+    }
+  }
+  def hiercat(e: Expression, dt: Type): Expression = {
+    dt match {
+      case t:VectorType => DoPrim(PrimOps.Cat, (0 to t.size).map(i =>
+          hiercat(WSubIndex(e, i, t.tpe, UNKNOWNGENDER),t.tpe)), Seq.empty[BigInt], UIntType(UnknownWidth))
+      case t:BundleType => DoPrim(PrimOps.Cat, t.fields.map(f =>
+          hiercat(WSubField(e, f.name, f.tpe, UNKNOWNGENDER), f.tpe)), Seq.empty[BigInt], UIntType(UnknownWidth))
+      case t:UIntType => e
+      case t:SIntType => e
+      case t => error("Unknown type encountered in toBits!")
+    }
+  }
+}
+
+object toBitMask {
+  def apply(e: Expression, dataType: Type): Expression = {
+    e match {
+      case ex: WRef => hiermask(ex,ex.tpe,dataType)
+      case ex: WSubField => hiermask(ex,ex.tpe,dataType)
+      case ex: WSubIndex => hiermask(ex,ex.tpe,dataType)
+      case t => error("Invalid operand expression for toBits!")
+    }
+  }
+  def hiermask(e: Expression, maskType: Type, dataType: Type): Expression = {
+    (maskType, dataType) match {
+      case (mt:VectorType, dt:VectorType) => DoPrim(PrimOps.Cat, (0 to mt.size).map(i =>
+          hiermask(WSubIndex(e, i, mt.tpe, UNKNOWNGENDER), mt.tpe, dt.tpe)), Seq.empty[BigInt], UIntType(UnknownWidth))
+      case (mt:BundleType, dt:BundleType) => DoPrim(PrimOps.Cat, (mt.fields zip dt.fields).map{ case (mf,df) =>
+        hiermask(WSubField(e, mf.name, mf.tpe, UNKNOWNGENDER), mf.tpe, df.tpe) }, Seq.empty[BigInt], UIntType(UnknownWidth))
+      case (mt:UIntType, dt) => groundBitMask(e, dt)
+      case (mt, dt) => error("Invalid type for mask component!")
+    }
+  }
+  def groundBitMask(e: Expression, dataType: Type): Expression = {
+    dataType match {
+      case dt:UIntType => DoPrim(PrimOps.Cat, List.fill(bitWidth(dt).intValue)(e), Seq.empty[BigInt], UIntType(UnknownWidth))
+      case dt:SIntType => DoPrim(PrimOps.Cat, List.fill(bitWidth(dt).intValue)(e), Seq.empty[BigInt], UIntType(UnknownWidth))
+      case dt => error("Mask type does not properly correspond with data type!")
+    }
+  }
+}
+
+object bitWidth {
+  def apply(dt: Type): BigInt = widthOf(dt)
+  def widthOf(dt: Type): BigInt = {
+    dt match {
+      case t:VectorType => t.size * bitWidth(t.tpe)
+      case t:BundleType => t.fields.map(f => bitWidth(f.tpe)).foldLeft(BigInt(0))(_+_)
+      case t:UIntType => baseWidth(t.width)
+      case t:SIntType => baseWidth(t.width)
+      case t => error("Unknown type encountered in bitWidth!")
+    }
+  }
+  def baseWidth(w: Width): BigInt = {
+    w match {
+      case w:IntWidth => w.width
```

Comment:
Another idiom of scala is the break out the case class, as it can sometimes result in cleaner-looking code:

```
w match {
   case IntWidth(width) => width
   ...
}
```